lib/spack/docs/tutorial_advanced_packaging.rst

```diff
@@ -0,0 +1,560 @@
+way to differentiate their answer to the query.
+
+---------------------------
+Package's build environment
+---------------------------
+
+Besides Spec's build interface, Spack provides means to set environment
+variables, either for yourself or for your dependent packages, and to
+attach attributes to your dependents. We'll see them next with the help
+of a few real use cases.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Set variables at build-time for yourself
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A few packages require their users to set some environment
+variables in order to be configured and built correctly. Spack
+permits you to accomplish this using the
+:py:func:`setup_environment <spack.package.PackageBase.setup_environment>` function.
+Let's try to see how it works by completing the ``elpa`` package:
+
+.. code-block:: console
+
+  root@advanced-packaging-tutorial:/# spack edit elpa
+
+In the end your method should look like:
+
+.. code-block:: python
+
+  def setup_environment(self, spack_env, run_env):
+      spec = self.spec
+
+      spack_env.set('CC', spec['mpi'].mpicc)
+      spack_env.set('FC', spec['mpi'].mpifc)
+      spack_env.set('CXX', spec['mpi'].mpicxx)
+      spack_env.set('SCALAPACK_LDFLAGS', spec['scalapack'].libs.joined())
+
+      spack_env.append_flags('LDFLAGS', spec['lapack'].libs.search_flags)
+      spack_env.append_flags('LIBS', spec['lapack'].libs.link_flags)
+
+The two arguments, ``spack_env`` and ``run_env``, are both instances of
+:py:class:`EnvironmentModifications <spack.environment.EnvironmentModifications>` and
+permit you to register modifications to either the build-time or the run-time
+environment of the package.
+At this point it's possible to proceed with the installation of ``elpa``:
+
+.. code-block:: console
+
+  root@advanced-packaging-tutorial:/# spack install elpa
+  ==> pkg-config is already installed in /usr/local/opt/spack/linux-ubuntu16.04-x86_64/gcc-5.4.0/pkg-config-0.29.2-ae2hwm7q57byfbxtymts55xppqwk7ecj
+  ==> ncurses is already installed in /usr/local/opt/spack/linux-ubuntu16.04-x86_64/gcc-5.4.0/ncurses-6.0-ukq4tccptm2rxd56d2bumqthnpcjzlez
+  ...
+  ==> Executing phase: 'build'
+  ==> Executing phase: 'install'
+  ==> Successfully installed elpa
+    Fetch: 3.94s.  Build: 41.93s.  Total: 45.87s.
+  [+] /usr/local/opt/spack/linux-ubuntu16.04-x86_64/gcc-5.4.0/elpa-2016.05.004-sdbfhwcexg7s2zqf52vssb762ocvklbu
+
+If you had modifications to ``run_env``, those would have appeared e.g. in the module files
+generated for the package.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Inject variables at build-time to your dependencies
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Another common occurrence for packages that support extensions, like ``r``
+or ``python``, or for packages that provide build tools is to
+require *their dependents* to have some environment variables set.
+
+The situation, even in this case, is similar to the one encountered previously.
```

Comment:
"The mechanism is similar to the what we just saw, except that we override the `setup_dependent_environment` function, which takes one additional..."